### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by DLochmelis33**
See [PLAN-150](https://youtrack.jetbrains.com/agiles/153-4013/current?backlog&issue=PLAN-150)

### CODE REVIEW

File: core/src/commonMain/kotlin/org/jetbrains/litmuskt/LitmusResult.kt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    for (stat in this) when (stat.type) {
+        LitmusOutcomeType.FORBIDDEN -> return LitmusOutcomeType.FORBIDDEN
+        LitmusOutcomeType.INTERESTING -> isInteresting = true
+        LitmusOutcomeType.ACCEPTED -> {} // ignore
+    }
+    return if (isInteresting) LitmusOutcomeType.INTERESTING else LitmusOutcomeType.ACCEPTED
+}
+
+fun List<LitmusResult>.mergeResults(): LitmusResult {
+    data class LTOutcomeStatTempData(var count: Long, val type: LitmusOutcomeType)
```
Comment (eupp) [inline core/src/commonMain/kotlin/org/jetbrains/litmuskt/LitmusResult.kt]: Let's rename into `LitmusOutcomeStatsAccumulator`

File: core/src/commonMain/kotlin/org/jetbrains/litmuskt/LitmusResult.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    val statMap = mutableMapOf<LitmusOutcome, LTOutcomeStatTempData>()
+    for (stat in this.flatten()) {
+        val tempData = statMap.getOrPut(stat.outcome) { LTOutcomeStatTempData(0L, stat.type) }
+        if (tempData.type != stat.type) error("merging conflicting stats: ${stat.outcome} is both ${stat.type} and ${tempData.type}")
```
Comment (eupp) [inline core/src/commonMain/kotlin/org/jetbrains/litmuskt/LitmusResult.kt]: Please put the `error` on a separate line, this line is too long to read.

File: core/src/commonMain/kotlin/org/jetbrains/litmuskt/ThreadlikeRunner.kt [hunk 3]
```diff
@@ -5,17 +5,22 @@ abstract class ThreadlikeRunner : LitmusRunner() {
     protected abstract fun threadlikeProducer(): Threadlike
 
     private fun <S : Any> threadFunction(threadContext: ThreadContext<S>) = with(threadContext) {
+        val testFunction = test.threadFunctions[threadIndex]
         for (i in states.indices) {
             if (i % syncPeriod == 0) barrier.await()
-            states[i].testThreadFunction()
+            states[i].testFunction()
         }
+        rangeResult = calcStats(states.view(resultCalcRange).asIterable(), test.outcomeSpec, test.outcomeFinalizer)
```
Comment (eupp) [inline core/src/commonMain/kotlin/org/jetbrains/litmuskt/ThreadlikeRunner.kt]: Could you please add a comment here explaining that as an optimization, each thread takes a sublist of all the results, and calculates the statistics for this sublist in parallel to other threads.

File: core/src/commonMain/kotlin/org/jetbrains/litmuskt/Utils.kt [hunk 4]
```diff
@@ -27,3 +27,25 @@ expect fun cpuCount(): Int
...
+ * Split a range into [n] parts of equal (+/- 1) length.
+ */
+fun IntRange.splitEqual(n: Int): List<IntRange> {
+    val size = endInclusive - start + 1
+    val len = size / n // base length of each sub-range
+    val remainder = size % n
+    val delim = start + (len + 1) * remainder // delimiter between lengths (l+1) and l
+    return List(n) { i ->
+        if (i < remainder) {
+            (start + i * (len + 1))..<(start + (i + 1) * (len + 1))
```
Comment (eupp) [inline core/src/commonMain/kotlin/org/jetbrains/litmuskt/Utils.kt:48]: Perhaps adding spaces around `..<` will improve readability

**Review by eupp — CHANGES_REQUESTED** [review]
There are couple of minor comments, otherwise it's a good PR, thanks!

File: core/src/commonMain/kotlin/org/jetbrains/litmuskt/Utils.kt (hunk 4 above)
Comment (DLochmelis33) [reply core/src/commonMain/kotlin/org/jetbrains/litmuskt/Utils.kt:48]: No spaces is (unfortunately) official codestyle

**Pull request merged**